feat: add ecdsa p-384 support

shengjiang3 opened the pull request:
GDC (Google Distributed Cloud) needs to support ECDSA-P384 keys for compliance. This change creates an EsSigner and EsVerifier class that is capable of supporting both ECDSA-P256 and ECDSA-P384 keys for backwards compatibility. The EsSigner and EsVerifier classes are plumbed through to the GDC service accounts and are used to both sign and verify JWTs.

This implementation was successfully tested against a GDC instance using both ECDSA-P256 and ECDSA-P384 keys.

shengjiang3 committed:
feat: add ecdsa p-384 support

GDC (Google Distributed Cloud) needs to support ECDSA-P384 keys for
compliance. This change creates an EsSigner and EsVerifier class that is
capable of supporting both ECDSA-P256 and ECDSA-P384 keys for backwards
compatibility. The EsSigner and EsVerifier classes are plumbed through
to the GDC service accounts and are used to both sign and verify JWTs.

This implementation was successfully tested against a GDC instance using
both ECDSA-P256 and ECDSA-P384 keys.

daniel-sanche committed:
Merge branch 'main' into gdc-es384

daniel-sanche commented on google/auth/crypt/__init__.py:
```diff
@@ -65,9 +67,11 @@
-    ES256Verifier = es256.ES256Verifier
+if es is not None:  # pragma: NO COVER
+    ES256Signer = es.EsSigner
+    ES256Verifier = es.EsVerifier
```
This is confusing to me. We do have `ES256Verifier`/`ES256Signer` as separate classes in the the es256 module, but instead you're importing the base class with a new name?

I understand they are both set up as aliases, but maybe they're not both needed?

daniel-sanche replied:
Yes, we need to keep them around to avoid breaking existing code. Can we replace these with imports from the `es256.py` file, so we are defining the alias in one place?

daniel-sanche commented on google/auth/crypt/es.py:
```diff
@@ -0,0 +1,208 @@
+# Copyright 2017 Google Inc.
```
Note to self (and other reviewers): most of this file is copied over from `es256.py`, so I only focused on the changes there

daniel-sanche commented on google/auth/crypt/es.py:
```diff
@@ -0,0 +1,208 @@
+        else:
+            self._r_s_size = 32
+            self._sha_algo = hashes.SHA256()
+            self._alg = "ES256"
```
These attributes are hard-coded in multiple places. Maybe it would be good to create a dataclass or named tuple to hold these? Something like:

```
@dataclass
class ESAttrs:
    rs_size: int
    sha_algo: hashes.HashAlgorithm
    name: str

    @classmethod
    def from_key(cls, key):
        if isinstance(private_key.curve, ec.SECP384R1):
            return cls(48, hashes.SHA384(), "ES384")
        else:
            # default to ES256
            return cls(32, hashes.SHA256(), "ES256")

class EsSigner:
    def __init__(self, private_key, key_id=None):
        self._es_attrs = ESAttrs.from_key(private_key)
```

shengjiang3 committed:
chore: Update es based on PR comments

daniel-sanche commented on google/auth/crypt/__init__.py:
```diff
@@ -41,13 +41,15 @@
 try:
-    from google.auth.crypt import es256
+    from google.auth.crypt import es
 except ImportError:  # pragma: NO COVER
```
IIRC, you can unimport a package using something like `del cryptography`. That should let you define a test method that doesn't have cryptography active. Did you try doing something like that?

daniel-sanche commented on google/auth/crypt/es.py:
```diff
@@ -0,0 +1,226 @@
+        self._key_id = key_id
+        self._attributes = ESAttributes.from_private_key(private_key)
+
+    @property  # type: ignore
```
why do we need the type: ignore here?

shengjiang3 replied:
Removed.

shengjiang3 committed:
chore: update based on PR feedback

- Changed ESAttributes to _ESAttributes.
- Merged from_private_key and from_public_key to from_key.

shengjiang3 commented on google/auth/crypt/es.py:
```diff
@@ -0,0 +1,226 @@
+
+
+@dataclass
+class ESAttributes:
```
Done.

shengjiang3 committed:
chore: use alias from es256

daniel-sanche reviewed (dismissed):
LGTM

daniel-sanche committed:
Merge branch 'main' into gdc-es384

shengjiang3 committed:
chore: Fix unit test failure

shengjiang3 committed:
chore: add code coverage

shengjiang3 committed:
chore: fix lint errors

shengjiang3 committed:
chore: fix one more lint error

daniel-sanche committed:
Merge branch 'main' into gdc-es384
